### ISSUE TITLE
Url Rewrite and Mod Rewrite Maps

### ISSUE THREAD

**Issue by ZestyBread** [aspnet/BasicMiddleware#55]
Both `RewriteMap` in mod_rewrite and `rewriteMap` in UrlRewrite are not supported in our rewrite module. From the initial scraping of examples, both of these features were rarely used, but it would be a nice addition.
From apache:
> The RewriteMap directive defines an external function which can be called in the context of RewriteRule or RewriteCond directives to perform rewriting that is too complicated, or too specialized to be performed just by regular expressions. The source of this lookup can be any of the types listed in the sections below, and enumerated in the RewriteMap reference documentation.

Ex:
 `RewriteMap MapName MapType:MapSource`

From IIS:
>A rewrite map is an arbitrary collection of name-value pairs that can be used within rewrite rules to generate the substitution URL during rewriting. Rewrite maps are particularly useful when you have a large set of rewrite rules and all of these rules use static strings (that is, when there is no pattern matching used). In those cases, instead of defining a large set of simple rewrite rules, you can put all the mappings into the rewrite map as keys and values between the input URL and the substitution URL. Then, to look up the substitution URL based on the input URL, you will have one rewrite rule that references the rewrite map.

```
<rewriteMap name="MyRewriteMap" defaultValue="">
    <add key="a.html" value="b.html" />
</rewriteMap>
```

**Comment by stewart-noll-q2** [comment]
@jkotalik do you happen to know if there's an eta on this.  Also, a nice-to-have: if a rewritemap provider interface is exposed and pluggable we could easily extend this to support things like db lookup, etc http://www.iis.net/learn/extensions/url-rewrite-module/using-custom-rewrite-providers-with-url-rewrite-module

**Comment by Tratcher** [comment]
There's no ETA for this feature, we're still gauging interest in it.

**Comment by davidpeden3** [comment]
@Tratcher how are you guys gauging interest? where can i go cast my vote?

**Comment by Tratcher** [comment]
@davidpeden3 Discussion here is sufficient

**Comment by eswann** [comment]
I'm very interested in this feature as well.

**Comment by davidpeden3** [comment]
@Tratcher i've got this feature implemented for IIS but am now blocked by https://github.com/aspnet/BasicMiddleware/issues/59 so i'm going to tackle that next. i'll submit a PR for both once they're in a good spot.

**Comment by Tratcher** [comment]
@davidpeden3 interesting. I recommend keeping the PRs separate to simplify the review process.

**Comment by davidpeden3** [comment]
@Tratcher certainly. i have global rules implemented now as well but am struggling with dependency issues in the test project so i'm currently unable to execute tests (which means that i would be writing them blind) and having to resort to manual testing to validate my code. i may give the 1.1.0 preview assemblies a try to see if that resolves my issue.

**Comment by Tratcher** [comment]
Yes running tests in VS with pre-release runtimes can be difficult to set up. Our command line build scripts set up a self-contained environment and can run the tests more reliably.

**Comment by davidpeden3** [comment]
@Tratcher i've got everything working as far as rewrite maps and global rules are concerned but realized that i need to actually proxy the request as it's not going anywhere. i found https://github.com/aspnet/Proxy but it looks like it can't be used as a reverse proxy because it's [hardcoded to a static host and port](https://github.com/aspnet/Proxy/blob/dev/src/Microsoft.AspNetCore.Proxy/ProxyMiddleware.cs#L83-L84). rather than messing with this repo, too, i think i'm gonna just write my own reverse proxy middleware that will do basically the same thing but use the rewritten uri.

am i on the right track here or am i missing something obvious in this rewrite middleware that handles proxying the rewritten requests?

**Comment by Tratcher** [comment]
Wait, what requires forwarding? global rules? I don't see that in the description, only that it mutates the full url. (let's take the global rules discussion over to #59)
